server-common/src/main/java/org/apache/kafka/server/util/FutureUtils.java

```diff
@@ -125,4 +127,21 @@ public static <T> CompletableFuture<T> combineFutures(
             return res;
         });
     }
+
+    /**
+     * Applies the given exception handler to all the futures provided in the list
+     * and returns a new list of futures.
+     *
+     * @param futures   A list of futures.
+     * @param fn        A function taking an exception to handle it.
+     * @return A list of futures.
+     */
+    public static <T> List<CompletableFuture<T>> mapExceptionally(
```

Comment:
This functionality is good idea. Maybe we can merge this it into `combineFutures` to make `combineFutures` to generate array with exception handler. For example:
```java
  public static <T> CompletableFuture<T> combineFutures(
      List<CompletableFuture<T>> futures,
      Supplier<T> init,
      BiConsumer<T, T> add
  ) {
    return combineFutures(futures, init, add, null);
  }

  public static <T> CompletableFuture<T> combineFutures(
      List<CompletableFuture<T>> futures,
      Supplier<T> init,
      BiConsumer<T, T> add,
      Function<Throwable, ? extends T> fn
  ) {
    final CompletableFuture<Void> allFutures = fn == null ? CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
      : CompletableFuture.allOf(futures.stream().map(f -> f.exceptionally(fn)).toArray(CompletableFuture[]::new));
    return allFutures.thenApply(v -> {
      final T res = init.get();
      futures.forEach(future -> add.accept(res, future.join()));
      return res;
    });
  }
```

The benefit is that we save a collection re-creation, and also `combineFutures` gets more functionality with acceptable complex ( 4 parameters )

Reply:
Interesting idea. I would prefer to keep `combineFutures` simple. `mapExceptionally` and `combineFutures` are two different things in my opinion. I also envision that we could have a need for something different that `exceptionally` in the future. If we go on that path, it means that we would need to overload `combineFutures` to support different ones and it will get messy.

Reply:
That makes sense to me. We can leave some rooms for refactor if there will be more use cases in the future. Let ship it instead of being over-engineering :)